core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -1332,10 +1334,14 @@ public RexNode makeIn(RexNode arg, List<? extends RexNode> ranges) {
     if (areAssignable(arg, ranges)) {
       final Sarg sarg = toSarg(Comparable.class, ranges, RexUnknownAs.UNKNOWN);
       if (sarg != null) {
-        final RexNode range0 = ranges.get(0);
+        final List<RelDataType> types = Util.transform(ranges, RexNode::getType);
+        final List<RelDataType> distinctTypes = Util.distinctList(types);
+        final RelDataType sargType = requireNonNull(
+            typeFactory.leastRestrictive(distinctTypes),
+            () -> "Can't find leastRestrictive type among " + distinctTypes);
         return makeCall(SqlStdOperatorTable.SEARCH,
             arg,
```

Comment:
`leastRestrictive` => `least restrictive`

Reply:
Is this the right fix ? It seems that the code does some logic for implicit type coercion here, and there is a breaking change:
Before the change, the in-compatible values in `IN` clause does not throw while with this patch it does, we need to confirm the behavior.

If we want to add implicit type coercion here, we better unify the logic with what `TypeCoercion` does.

Reply:
`leastRestrictive => least restrictive`
The exception message is keep consistent with `RexSargBuilder#RexSargBuilder` in `RexSimplify`.
I think it's a better user experience for user.

Reply:
> It seems that the code does some logic for implicit type coercion here
The pr does not aim to fix implicit type coercion, it only want to solve the behavior of `RexBulder#makeIn` if the arguments are compatibule, just like discussion in the JIRA:
![image](https://user-images.githubusercontent.com/1525333/143808516-287f6065-f46d-479a-a897-abf5b4639aea.png).
I could create another JIRA to solve implicit type coercion.
Let's solve the problem one by one.
> Before the change, the in-compatible values in IN clause does not throw while with this patch it does, we need to confirm the behavior.
The old behavior is not correct which we need to fix.